events/events_test.go

```diff
@@ -1,44 +1,267 @@
+	sscanner := bufio.NewScanner(sreader)
+	expectedEvent := "status_update_event"
+	// when
+	event, _ := ParseEvent(sscanner)
+	// then
+	assert.Equal(t, expectedEvent, event.Type)
+	assert.Equal(t, []byte("testEventData\n"), event.Body)
 }
 
-func TestMissingTimestamp(t *testing.T) {
+func TestScanLinesWhenSomeDataWithNewLineIsProvided(t *testing.T) {
 	t.Parallel()
-	out, err := ParseEvent([]byte(`{"eventType":"test"}`))
-	assert.EqualError(t, err, "Missing timestamp")
-	assert.Equal(t, out, Event{})
+	// given
+	input := []byte("abcd\n")
+	// when
+	advance, token, err := ScanLines(input, false)
+	require.NoError(t, err)
+	expectedAdvance := 5
+	expectedToken := []byte("abcd")
+	// then
+	assert.Equal(t, expectedAdvance, advance)
+	assert.Equal(t, expectedToken, token)
 }
 
-func TestEventTypeInvalidJson(t *testing.T) {
+func TestScanLinesWhenSomeDataWithCarriageReturnIsProvided(t *testing.T) {
 	t.Parallel()
+	// given
+	input := []byte("abcd\r")
+	// when
+	advance, token, err := ScanLines(input, false)
+	require.NoError(t, err)
+	expectedAdvance := 5
+	expectedToken := []byte("abcd")
+	// then
+	assert.Equal(t, expectedAdvance, advance)
+	assert.Equal(t, expectedToken, token)
+}
 
-	out, err := ParseEvent([]byte(`not a json`))
-	assert.Error(t, err)
-	assert.Equal(t, out, Event{})
+func TestScanLinesWhenSomeDataWithCarriageReturnAndNewLineIsProvided(t *testing.T) {
+	t.Parallel()
+	// given
+	input := []byte("abcd\r\n")
+	// when
+	advance, token, err := ScanLines(input, false)
+	require.NoError(t, err)
+	expectedAdvance := 6
+	expectedToken := []byte("abcd")
+	// then
+	assert.Equal(t, string(expectedAdvance), string(advance))
+	assert.Equal(t, expectedToken, token)
+}
+
+func TestScanLinesWhenSomeDataWithoutCarriageReturnAndNewLineIsProvided(t *testing.T) {
+	t.Parallel()
+	// given
+	input := []byte("abcd")
+	// when
+	advance, token, err := ScanLines(input, false)
+	require.NoError(t, err)
+	expectedAdvance := 0
+	expectedToken := []byte(nil)
+	// then
+	assert.Equal(t, expectedAdvance, advance)
+	assert.Equal(t, expectedToken, token)
+}
+
+func TestScanLinesWhenSomeDataWithoutCarriageReturnAndNewLineIsProvidedAndFunctionIsAtEOF(t *testing.T) {
+	t.Parallel()
+	// given
+	input := []byte("abcd")
+	// when
+	advance, token, err := ScanLines(input, true)
+	require.NoError(t, err)
+	expectedAdvance := 4
+	// TODO(tz) consider returning passed string,
+	// question here is how Scan function will handle it
+	expectedToken := []byte("abcd")
+	// then
+	assert.Equal(t, expectedAdvance, advance)
+	assert.Equal(t, expectedToken, token)
+}
+
+func TestScanLinesWhenSomeDataWithoutCarriageReturnAndNewLineIsProvidedAndFunctionIsAtEOFAndLFIsAtTheEnd(t *testing.T) {
+	t.Parallel()
+	// given
+	input := []byte("abcd\n")
+	// when
+	advance, token, err := ScanLines(input, true)
+	require.NoError(t, err)
+	expectedAdvance := 5
+	expectedToken := []byte("abcd")
+	// then
+	assert.Equal(t, expectedAdvance, advance)
+	assert.Equal(t, expectedToken, token)
```

Comment:
Most of this tests could be merged into one with different test cases run as a subtests.